src/java.base/share/classes/javax/crypto/KEM.java

```diff
@@ -0,0 +1,731 @@
+         * @return the secret key
+         */
+        public SecretKey key() {
+            return key;
+        }
+
+        /**
+         * Returns the key encapsulation message.
+         *
+         * @return the key encapsulation message. A new copy of the byte array
+         *      is returned.
+         */
+        public byte[] encapsulation() {
+            return encapsulation.clone();
+        }
+
+        /**
+         * Returns the optional parameters in a byte array.
+         *
+         * @return the optional parameters in a byte array or {@code null}
+         *      if not specified. A new copy of the byte array is returned.
+         */
+        public byte[] params() {
+            return params == null ? null : params.clone();
+        }
+    }
+
+    /**
+     * An encapsulator, generated by {@link #newEncapsulator} on the KEM
+     * sender side.
+     * <p>
+     * This class represents the key encapsulation function of a KEM.
+     * Each invocation of the {@code encapsulate} method generates a
+     * new secret key and key encapsulation message that is returned
+     * in an {@link Encapsulated} object.
+     *
+     * @since 21
+     */
+    public static final class Encapsulator {
+
+        private final KEMSpi.EncapsulatorSpi e;
+        private final Provider p;
+
+        private Encapsulator(KEMSpi.EncapsulatorSpi e, Provider p) {
+            assert e != null;
+            assert p != null;
+            this.e = e;
+            this.p = p;
+        }
+
+        /**
+         * Returns the name of the provider.
+         *
+         * @return the name of the provider
+         */
+        public String providerName() {
+            return p.getName();
+        }
+
+        /**
+         * The key encapsulation function.
+         * <p>
+         * This method is equivalent to
+         * {@code encapsulate(0, secretSize(), "Generic")}. This combination
+         * of arguments must be supported by every implementation.
+         * <p>
+         * The generated secret key is usually passed to a key derivation
+         * function (KDF) as the input keying material.
+         *
+         * @return a {@link Encapsulated} object containing the shared
+         *          secret, key encapsulation message, and optional parameters.
+         *          The shared secret is a {@code SecretKey} containing all of
+         *          the bytes of the secret, and an algorithm name of "Generic".
+         */
+        public Encapsulated encapsulate() {
+            return encapsulate(0, secretSize(), "Generic");
+        }
+
+        /**
+         * The key encapsulation function.
+         * <p>
+         * Each invocation of this method generates a new secret key and key
+         * encapsulation message that is returned in an {@link Encapsulated} object.
+         * <p>
+         * An implementation may choose to not support arbitrary combinations
+         * of {@code from}, {@code to}, and {@code algorithm}.
```

Comment:
As previously discussed, I think having a code example of the `from` and `to` would be good idea.  That way it's very clear going from 0 to 32 is 32 bytes and not 33.  And an example would be a good idea in the SPI.